### DISCUSSION TITLE
Why `newExpr` requires `new` keyword?

### DISCUSSION THREAD

**Discussion by taichi-ishitani** [apple/pkl#419] [Q&A]
I have a question about `newExpr`.

I'm using `YAML` as the input format of my app like this and tring to rewrite it by using Pkl.
https://github.com/rggen/rggen-sample/blob/master/uart_csr.yml

This is the result.
```pkl
import "rggen.pkl"

register_block = (rggen.register_block) {
  name = "uart_csr"
  byte_size = 32

  (rggen.register) {
    name            = "rbr"
    offset_address  = 0x00
    type            = new { "indirect"; new { "lcr.dlab"; 0 } }
    comment         = "Receiver Buffer Register"

    (rggen.bit_field) {
      bit_assignment  = new { lsb = 0; width = 8 }
      type            = "rotrg"
    }
  }
}
```

This trial led me to think that `{}` blocks without new keyword looks better.
Is there any reason why the `newExpr` requires the `new` keyword?

**Comment by odenix** [comment]
Most of the `new` keywords in your code shouldn't be required. Instead of `foo = new { ... }`, you can write `foo { ... }`.

**Comment by taichi-ishitani** [comment]
Yes, I know but I'd like to use the same assignment notation like below.

```pkl
foo = 1
bar = { baz = 1 }
```

In addition, an `newExpr` within a `{}` block requires the `new` keyword. I'd like to ommit this `new` keyword too.

```pkl
foo { 0 new { bar = 1 } } // Need new keyword
baz { 2 { qux = 3 } } // I'd like to write like this
```

**Reply by odenix** [reply]
Then what would the following mean? `baz { x { qux = 3 } }`

**Reply by taichi-ishitani** [reply]
I think this case should be interpreted to

```pkl
baz = new { x; new { qux = 3 } }
```

**Reply by odenix** [reply]
But `x { … }` already has a different meaning, namely to configure the property named `x`.

**Reply by taichi-ishitani** [reply]
I understood.
This should interpreted to 

```pkl
baz = new { x = new { qux = 3 } }
```

If we want to treat `x` as a member reference we need to insert a semicolon after `x`.

```pkl
baz { x; { qux = 3 } }
```

**Reply by odenix** [reply]
I think omitting `new` creates more problems than it solves. But you can try to change the grammar.

Personally, I’d like to be able to write `(Person) {…}` instead of `new Person {…}`.

**Reply by bioball** [reply]
> Personally, I’d like to be able to write (Person) {…} instead of new Person {…}.

You actually _can_ write `(Person) { ... }`, but it's not preferred.

**Reply by odenix** [reply]
That's good to know!
I've been toying with a typed config format that is compatible with Pkl.
For a config format, a unified notation for non-primitive values would be very desirable, and IMO the `()` syntax reads much better:
```
x = (Person) {...}
y = (Listing) {...}
z = (Set) {...}
```

**Reply by holzensp** [reply]
You _can_ indeed do this, but as bioball said, it's not preferred. That is to say; it may well still go away. The problem is that it muddies the namespaces of types and values (which is bad for autocomplete). The idea that parentheses are an operator that takes a class and produces an empty instance to amend is confusion waiting to happen, if you ask me.

**Reply by bioball** [reply]
Another issue with this syntax is that it's not possible to instantiate more complex types, like `new Listing<Person>`.

**Reply by taichi-ishitani** [reply]
I think it would be nice if Pkl has `object literal` like other languages.
https://github.com/apple/pkl/discussions/419#discussioncomment-9091186
What do you think about this?

**Reply by odenix** [reply]
> Another issue with this syntax is that it's not possible to instantiate more complex types

This may be a concern for Pkl, but a typed config/data format only needs to know the type of each individual value. However, `new` looks strange in a data format.

> I think it would be nice if Pkl has object literal like other languages.

Pkl’s object literal syntax is `new Type {…}`. Only in some cases this can be abbreviated to `new {…}`.

If I could, I'd be tempted to phase out `new Type {...}` in favor of `(Type) {...}` and to only allow simple types in object literals.
Not sure what to do about `new {...}` though.

**Reply by holzensp** [reply]
The `(Type) { ... }` syntax is suboptimal in that it conflates types and values (bad for autocomplete) and it makes the meaning of parentheses significantly more loaded than what most people would assume (`(<expr>) { ... }` now means "evaluate the object from `<expr>` and amend it", whereas the `(Type) { ... }` instantiation means "evaluate the object from `<expr>` and, if it's a type, instantiate it, otherwise use it as-is, and then amend it").

**Comment by bioball** [comment]
In Pkl, object bodies always amend _something_. `new {}` amends the inferred default value. For example, consider [listing defaults](https://pkl-lang.org/main/current/language-reference/index.html#default-element):

```groovy
class Person {
  name: String
  birthday: String
}

people = new Listing<Person> {
  default { name = "Ellis" }
  new { birthday = "2020/01/01" }
}
```

Here, the `new` is derived from the default value defined right above it (you end up with `new Person { name = "Ellis"; birthday = "2020/01/01" }`)

Simply using `{}` for notation would perhaps make this behavior a little too hidden. This feels too surprising to me:

```groovy
people = new Listing<Person> {
  default { name = "Ellis" };
  { birthday = "2020/01/01" }
}
```

There's also a grammar ambiguity here; this is parsed as a chained amends expression `(foo) {} {} {} {}`. If `{}` can also mean "standalone object", then we have now introduced an ambiguity.

```groovy
myObj {
  (foo) {}
  {}
  {}
  {}
}
```

**Reply by taichi-ishitani** [reply]
I understood it's difficult to omit `new` keyword simply and new notation is needed for the `newExpr` without `new` keyword.
How about using a quotation mark instead of `new` keyword like this?

```pkl
foo = '{ bar = 0 }
```

BTW, how can I highlight Pkl code in Markdown?

**Reply by bioball** [reply]
It's certainly possible to change the grammar. This is subjective, but I think this is more legible:

```
foo = new { bar = 0 }
```

This seems less intuitive to me:

```
foo = '{ bar = 0 }
```

I'm using the Groovy syntax highlighter right now for Pkl code. That seems to be the best option for now, until https://github.com/github-linguist/linguist/pull/6730 gets merged

````
```groovy
foo = new {}
```
````

**Reply by taichi-ishitani** [reply]
Most langurages have array literal/hash literal. This is Ruby example.

```ruby
a = [0, 1, 2]
h = { foo: 0, bar: 1, baz: 2 }
```

I would like a similar syntax for Pkl.

> I'm using the Groovy syntax highlighter right now for Pkl code. 

Thanks! I will use this technique.
